client/common/Button.tsx

```diff
@@ -1,24 +1,94 @@
 import React from 'react';
-import PropTypes from 'prop-types';
 import styled from 'styled-components';
-import { Link } from 'react-router-dom';
-
+import { Link, LinkProps } from 'react-router-dom';
 import { remSize, prop } from '../theme';
 
 const kinds = {
   primary: 'primary',
   secondary: 'secondary'
-};
+} as const;
 
 const displays = {
   block: 'block',
   inline: 'inline'
+} as const;
+
+const buttonTypes = {
+  button: 'button',
+  submit: 'submit'
+} as const;
+
+type Kind = keyof typeof kinds;
+type Display = keyof typeof displays;
+type ButtonType = keyof typeof buttonTypes;
+
+type StyledButtonProps = {
+  kind: Kind;
+  display: Display;
+};
+
+type SharedButtonProps = {
+  /**
+   * The visible part of the button, telling the user what
+   * the action is
+   */
+  children?: React.ReactNode;
+  /**
+      If the button can be activated or not
+    */
+  disabled?: boolean;
+  /**
+   * The display type of the button—inline or block
+   */
+  display?: Display;
+  /**
+   * SVG icon to place after child content
+   */
+  iconAfter?: React.ReactNode;
+  /**
+   * SVG icon to place before child content
+   */
+  iconBefore?: React.ReactNode;
+  /**
+   * If the button content is only an SVG icon
+   */
+  iconOnly?: boolean;
+  /**
+   * The kind of button - determines how it appears visually
+   */
+  kind?: Kind;
+  /**
+   * Specifying an href will use an <a> to link to the URL
+   */
+  href?: string | null;
```

Comment:
feels like if href can be undefined then we should avoid also letting it be null?